src/go/rpk/pkg/cli/generate/app.go

```diff
@@ -0,0 +1,430 @@
+		s := strings.SplitN(saslCred, ":", 2)
+		if len(s) < 2 {
+			return "", "", "", fmt.Errorf("unable to parse %q: credentials does not follow the <user>:<pass> format", saslCred)
+		}
+		user = s[0]
+		pass = s[1]
+		err = createUser(ctx, fs, p, user, pass)
+		if err != nil {
+			return "", "", "", err
+		}
+	} else {
+		if p.HasSASLCredentials() {
+			user = p.KafkaAPI.SASL.User
+			pass = p.KafkaAPI.SASL.Password
+		}
+	}
+	return
+}
+
+func createUser(ctx context.Context, fs afero.Fs, p *config.RpkProfile, user, pass string) error {
+	cl, err := adminapi.NewClient(fs, p)
+	if err != nil {
+		return fmt.Errorf("unable to initialize admin client: %v", err)
+	}
+	err = cl.CreateUser(ctx, user, pass, adminapi.ScramSha256)
+	if err != nil {
+		return fmt.Errorf("unable to create user %q: %v", user, err)
+	}
+	adm, err := kafka.NewAdmin(fs, p)
+	if err != nil {
+		return fmt.Errorf("unable to initialize kafka client: %v", err)
+	}
+	// Read-Write ACL:
+	b := kadm.NewACLs().
+		ResourcePatternType(kadm.ACLPatternLiteral).
+		Allow(user).
+		Clusters().
+		Topics("*").
+		Operations(kadm.OpRead, kadm.OpWrite, kadm.OpCreate)
+	b.PrefixUser()
+	_, err = adm.CreateACLs(ctx, b)
+	if err != nil {
+		return fmt.Errorf("unable to Create ACL for user %v: %v", user, err)
+	}
+	fmt.Printf("Successfully created user %q\n", user)
+	return nil
+}
+
+// extractTarGzAndReplace extracts and expand the sample app directory to destFS
+// and replaces the values with the passed tmplData. Returns the path where file
+// was extracted.
+func extractTarGzAndReplace(tplFS iofs.FS, destFS afero.Fs, data tmplData, file, out string) (string, error) {
+	r, err := tplFS.Open(file)
+	if err != nil {
+		return "", err
+	}
+	gr, err := gzip.NewReader(r)
+	if err != nil {
+		return "", err
+	}
+	tr := tar.NewReader(gr)
+	var extractPath string
+	for {
+		fh, err := tr.Next()
+		if errors.Is(err, io.EOF) {
+			break
+		}
+		if err != nil {
+			return "", err
+		}
+		switch fh.Typeflag {
+		case tar.TypeDir:
+			dir := fh.Name
+			if out != "" {
+				dir = filepath.Join(out, fh.Name)
+			}
+			if exists, _ := afero.DirExists(destFS, dir); !exists {
+				if err := destFS.MkdirAll(dir, 0o755); err != nil {
+					return "", err
+				}
+			}
+			// This is saved the first time only.
+			if extractPath == "" {
+				extractPath = dir
+			}
+		case tar.TypeReg:
+			// The file is a Go template, we parse it and copy it to the
+			// destination with the values replaced.
+			b, err := io.ReadAll(tr)
+			if err != nil {
+				return "", err
+			}
+			tpl, err := template.New("sample").Parse(string(b))
+			if err != nil {
+				return "", err
+			}
+			dest := fh.Name
+			if out != "" {
+				dest = filepath.Join(out, fh.Name)
+			}
+			temp, err := afero.TempFile(destFS, filepath.Dir(dest), "rpk-starter-app-*")
+			if err != nil {
+				return "", err
+			}
+			err = tpl.Execute(temp, data)
+			if err != nil {
+				return "", err
+			}
+			err = destFS.Rename(temp.Name(), dest)
+			if err != nil {
```

Comment:
I couldn't remove the TypeDir switch since `rpkos.ReplaceFile` not only creates the dir, but also writes the content, and when we are extracting the tar we want to create the dir but not do anything else.

The alternative (not sure either) will be `case tar.TypeDir: // do nothing` and let rpkos.ReplaceFile handle the dir creation. But I prefer leaving it as it is because it's explicit what we are doing.